design/one-pager-configuration-stacks.md

```diff
@@ -0,0 +1,282 @@
+abilites, to deploy an application and all other things that you can do with
+Kubernetes controllers since main components of stacks are CRDs and controllers
+watching them.
+
+To achieve the mentioned goals, we can publish the YAML files in a stack with a
+controller that has its own CRD with high level parameters. Controller deploys
+and updates the YAMLs that are packaged in the stack image according to user's
+input on the CR instance of that stack that user creates.
+
+As sky is the limit when it comes to stacks, we need to put down some general
+rules to be adhered by initial configuration stacks.
+
+* It should be possible to have different sets of configurations in one stack
+  and user should be able to choose from them via the CR that is an instance
+  of stack's CRD.
+  * Note that it's highly preferred to have only one custom resource definition that
+    the stack uses. You can expose the high level variations of the parameters
+    through `spec` instead of having different custom resource definitions in the
+    same stack or you can write another stack with its own CRD for that purpose.
+* A minimal high level set of configurations are exposed to the user through
+  the stack's CR, such as reference to the cloud credential secret or `region`.
+* All resources should be labelled referring to the CR instance of the stack.
+* In YAMLs, resources should refer to each other using cross-resource references
+  wherever possible.
+* Controller should update the resources continuously and treat the given CR as
+  the source of truth even though user manually changes the resources that are
+  deployed by the stack.
+* CRD of the configuration stack should be cluster-scoped since the author of
+  the CR instances of the configuration stack is assumed to be the infrastructure
+  owner.
+* Stack type will be `ClusterStack` as most of the resources it deploys are
+  cluster-scoped.
+
+## User Experience
+
+1. Install AWS stack and create a secret that contains AWS credentials.
+2. Install a AWS configuration stack via:
+```yaml
+# exact apiVersion is TBD.
+apiVersion: stacks.crossplane.io/v1alpha1
+kind: ClusterStackInstall
+metadata:
+  name: minimal-aws
+  namespace: crossplane-system
+spec:
+  package: "crossplane/minimal-aws:latest"
+```
+
+Now, I want an environment where all my database instances and kubernetes clusters
+are connected to each other in a private VPC, which what this specific configuration
+stack does. Create the following:
+
+```yaml
+apiVersion: aws.configurationstacks.crossplane.io/v1alpha1
+kind: MinimalAWS
+metadata:
+  name: small-infra
+spec:
+  credentialsSecretRef:
+    name: aws-credentials
+    namespace: crossplane-system
+```
+
+Then I wait for `Synced` condition to become `true`. After it's done, all resources
+are deployed.
+```yaml
+apiVersion: aws.configurationstacks.crossplane.io/v1alpha1
+kind: MinimalAWS
+metadata:
+  name: small-infra
+spec:
```

Comment:
We'll also want to include spec attributes like region, projectId, and likely allow overriding defaults for things like CIDR / IP ranges for nodes, pods, services, etc. If we're using kustomize then this would likely generate a patch.yaml similar to what @hasheddan put together [here](https://github.com/crossplaneio/crosscd/blob/master/classpacks/workflow/argo/infra/gcp/patch.yaml).

Reply:
The `patch.yaml` approach requires the types to be known by the controller and iteration over each and every one of them, which is not ideal IMO since the developer will have to implement more complex functions.

I've experimented with [transformer configs of Kustomize for variant referencing](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/transformerconfigs/crd/README.md#adding-a-custom-resource) and I think it works for us. An example implementation is here https://github.com/muvaf/minimal-gcp/pull/2 . As you'll see, we define the region as `$(REGION)` in the YAML files and then in the controller code, only [define the path in the CR](https://github.com/muvaf/minimal-gcp/pull/2/files#diff-1fa7bf35a53cd97938369f8733221602R70) that the value should be taken.